### PR TITLE
Feature/ ATR-134, ATR-88, ATR-182 

### PR DESCRIPTION
**Pull request by saloponov**

### CODE REVIEW

File: src/Acuminator/Acuminator.Analyzers/Resources.resx [hunk 1]
```diff
@@ -252,4 +252,10 @@
   <data name="PX1032Title" xml:space="preserve">
     <value>DAC properties cannot contain method invocations</value>
   </data>
+  <data name="PX1030Fix" xml:space="preserve">
+    <value>Replace with the PXUnboundDefaultAttribute or set PersistingCheck to Nothing</value>
```
Comment (saloponov) [inline src/Acuminator/Acuminator.Analyzers/Resources.resx]: @k-s-popova , please check this string.

File: src/Acuminator/Acuminator.Analyzers/Resources.resx [hunk 2]
```diff
@@ -252,4 +252,10 @@
...
+  </data>
+  <data name="PX1030Title" xml:space="preserve">
+    <value>Incorrect PXDefaultAttribute to exisiting field </value>
```
Comment (saloponov) [inline src/Acuminator/Acuminator.Analyzers/Resources.resx]: @k-s-popova , please check this string.

**Comment by SENya1990** [comment]
@saloponov please merge master into your branch and resolve merge conflicts

File: src/Acuminator/Acuminator.Utils/RoslynExtensions/PXFieldAttributes/AttributeInformation.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+		public bool IsBoundAttribute(ITypeSymbol attributeSymbol)
+		{
+			//Get this symbols from PXContext
+			//PX.Data.PXDBFieldAttribute
+			var dbFieldAttribute = _context.FieldAttributes.PXDBFieldAttribute;
+			return AttributeDerivedFromClass(attributeSymbol, dbFieldAttribute);
+		}
+		
+		public bool AreBoundAttributes(IEnumerable<ITypeSymbol> attributesSymbols)
```
Comment (SENya1990) [inline src/Acuminator/Acuminator.Utils/RoslynExtensions/PXFieldAttributes/AttributeInformation.cs]: @saloponov , this names is misleading. AreBoundAttributes implies check that every attribute in collection is "Bound" while you actually search for the first "Bound" attribute.

File: src/Acuminator/Acuminator.Utils/RoslynExtensions/PXFieldAttributes/AttributeInformation.cs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+			}
+			return false;
+		}
+
+		public bool IsBoundAttribute(ITypeSymbol attributeSymbol)
+		{
+			//Get this symbols from PXContext
+			//PX.Data.PXDBFieldAttribute
+			var dbFieldAttribute = _context.FieldAttributes.PXDBFieldAttribute;
+			return AttributeDerivedFromClass(attributeSymbol, dbFieldAttribute);
```
Comment (SENya1990) [inline src/Acuminator/Acuminator.Utils/RoslynExtensions/PXFieldAttributes/AttributeInformation.cs]: @saloponov  the check if attribute is derived from PXDBFieldAttribute can't be used to check if the attribute is DB bound. 
@dropsonic  correct me if I'm wrong here, but there are PXDBUserPasswordAttribute and PXDBCalcedAttribute which do not derive from the PXDBFieldAttribute and don't have any aggregator attributes in inheritance chain which are marked with PXDBFieldAttribute

File: src/Acuminator/Acuminator.Utils/RoslynExtensions/PXFieldAttributes/AttributeInformation.cs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+		public bool AttributeDerivedFromClass(ITypeSymbol attributeSymbol, ITypeSymbol type, int depth = 10)
+		{
+			attributeSymbol.ThrowOnNull(nameof(attributeSymbol));
+			type.ThrowOnNull(nameof(type));
+			if (depth <= 0 || depth > 100)
+				return false;
+			if (ContainsBaseType(attributeSymbol, type))
+				return true;
+
+			//get symbols from PXContext
```
Comment (SENya1990) [inline src/Acuminator/Acuminator.Utils/RoslynExtensions/PXFieldAttributes/AttributeInformation.cs]: @saloponov  this comment is redundant. It's already obvious from the code what is done here

File: src/Acuminator/Acuminator.Analyzers/Resources.resx (hunk 1 above)
Comment (k-s-popova) [reply src/Acuminator/Acuminator.Analyzers/Resources.resx]: @saloponov, change the message as follows: 
Replace with PXUnboundDefaultAttribute or set PersistingCheck to Nothing

File: src/Acuminator/Acuminator.Analyzers/Resources.resx (hunk 2 above)
Comment (k-s-popova) [reply src/Acuminator/Acuminator.Analyzers/Resources.resx]: @saloponov, I would change the message as follows:
The PXDefault attribute of the field is used incorrectly

File: src/Acuminator/Acuminator.Analyzers/Analyzers/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeAnalyzer.cs [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+		}
+
+
+		public static async Task<Location> GetAttributeLocationAsync(AttributeData attribute, CancellationToken cancellationToken)
+		{
+			SyntaxNode attributeSyntaxNode = null;
+
+			try
+			{
+				attributeSyntaxNode = await attribute.ApplicationSyntaxReference.GetSyntaxAsync(cancellationToken)
```
Comment (dropsonic) [inline src/Acuminator/Acuminator.Analyzers/Analyzers/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeAnalyzer.cs]: I suggest using `cancellationToken.ThrowIfCancellationRequested` inside `GetSyntaxAsync` and avoid try/catch here.

File: src/Acuminator/Acuminator.Analyzers/FixProviders/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeFix.cs [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+		public override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
+
+		public override Task RegisterCodeFixesAsync(CodeFixContext context)
+		{
+			return Task.Run(() =>
+			{
+				var diagnostic = context.Diagnostics.FirstOrDefault(d => d.Id == Descriptors.PX1030_DefaultAttibuteToExisitingRecords.Id);
+
+				if (diagnostic == null)
```
Comment (dropsonic) [inline src/Acuminator/Acuminator.Analyzers/FixProviders/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeFix.cs]: This code is unnecessary; this scenario is handled by VS - supported diagnostics are specified explicitly in `FixableDiagnosticIds`

File: src/Acuminator/Acuminator.Analyzers/FixProviders/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeFix.cs [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+	public class DacExtensionDefaultAttributeFix : CodeFixProvider
+	{
+		public override ImmutableArray<string> FixableDiagnosticIds { get; } =
+			ImmutableArray.Create(Descriptors.PX1030_DefaultAttibuteToExisitingRecords.Id);
+
+		public override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
+
+		public override Task RegisterCodeFixesAsync(CodeFixContext context)
+		{
+			return Task.Run(() =>
```
Comment (dropsonic) [inline src/Acuminator/Acuminator.Analyzers/FixProviders/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeFix.cs]: Are you sure that `Task.Run` is needed here? I don't see any CPU-consuming operations here. A lot of background tasks can lead to thread pool starvation.

File: src/Acuminator/Acuminator.Analyzers/FixProviders/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeFix.cs [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+		private async Task<Document> ReplaceIncorrectDefaultAttribute(Document document, TextSpan span, bool isBoundField,CancellationToken cancellationToken)
+		{
+			SyntaxNode root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+			 if (!(root?.FindNode(span) is AttributeSyntax attributeNode))
+				return document;
+			if (!(attributeNode.Parent is AttributeListSyntax attributeList))
+				return document;
+			if (cancellationToken.IsCancellationRequested)
```
Comment (dropsonic) [inline src/Acuminator/Acuminator.Analyzers/FixProviders/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeFix.cs]: Use `cancellationToken.ThrowIfCancellationRequested()`

File: src/Acuminator/Acuminator.Analyzers/FixProviders/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeFix.cs [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+				return document;
+
+			SyntaxGenerator generator = SyntaxGenerator.GetGenerator(document);
+			var memberAccessExpression = generator.MemberAccessExpression(generator.IdentifierName("PXPersistingCheck"), generator.IdentifierName("Nothing"));
```
Comment (dropsonic) [inline src/Acuminator/Acuminator.Analyzers/FixProviders/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeFix.cs]: Please move hardcoded names to constants

File: src/Acuminator/Acuminator.Analyzers/FixProviders/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeFix.cs [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+					{
+						//SyntaxNode newAttributeList = generator.AddAttributes(new AttributeListSyntax(), new SyntaxNode);
+						var newAttributeList = generator.AddAttributeArguments(attributeNode, new SyntaxNode[] { persistingAttributeArgument }) as AttributeListSyntax;
+						//var newAttributeList = attributeList.ReplaceNode(attributeNode, newAttributeNode);
+						modifiedRoot = root.ReplaceNode(attributeNode, newAttributeList.Attributes[0]);
+					}
+
+/*					SyntaxNode newAttributeNode = generator.InsertAttributeArguments(attributeNode, 1, new SyntaxNode[] { persistingAttributeArgument });
+					
+					//var newAttribute = generator.InsertAttributeArguments(attributeNode, 1, new SyntaxNode[] { persistingAttributeArgument });
```
Comment (dropsonic) [inline src/Acuminator/Acuminator.Analyzers/FixProviders/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeFix.cs]: commented code

File: src/Acuminator/Acuminator.Analyzers/FixProviders/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeFix.cs [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+						AttributeArgumentSyntax argument = persistingCheckArguments.First();
+						
+						persistingAttributeArgument = argument.ReplaceNode(argument.Expression, memberAccessExpression);
+						var newAttributeNode = attributeNode.ReplaceNode(argument, persistingAttributeArgument);
+						var newAttributeList = attributeList.ReplaceNode(attributeNode, newAttributeNode);
+						modifiedRoot = root.ReplaceNode(attributeList, newAttributeList);
+					}
+					else
+					{
+						//SyntaxNode newAttributeList = generator.AddAttributes(new AttributeListSyntax(), new SyntaxNode);
```
Comment (dropsonic) [inline src/Acuminator/Acuminator.Analyzers/FixProviders/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeFix.cs]: commented code

File: src/Acuminator/Acuminator.Analyzers/FixProviders/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeFix.cs [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+						var newAttributeList = generator.AddAttributeArguments(attributeNode, new SyntaxNode[] { persistingAttributeArgument }) as AttributeListSyntax;
+						//var newAttributeList = attributeList.ReplaceNode(attributeNode, newAttributeNode);
```
Comment (dropsonic) [inline src/Acuminator/Acuminator.Analyzers/FixProviders/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeFix.cs]: commented code

File: src/Acuminator/Acuminator.Tests/Tests/Unit/AttributeInformationTests.cs [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+			var properties = syntaxRoot.DescendantNodes().OfType<PropertyDeclarationSyntax>();
+
+			foreach (var property in properties)
+			{
+				var typeSymbol = semanticModel.GetDeclaredSymbol(property);
+				var attributes = typeSymbol.GetAttributes();
+				foreach (var attribute in attributes)
+				{
+					var attributeInformation = new AttributeInformation(pxContext);
+					var defaultAttribute = semanticModel.Compilation.GetTypeByMetadataName("PX.Data.PXDefaultAttribute");//rewrite to pxContext.PXDefaultAttribute
```
Comment (dropsonic) [inline src/Acuminator/Acuminator.Tests/Tests/Unit/AttributeInformationTests.cs]: Use `pxContext`

File: src/Acuminator/Acuminator.Utils/RoslynExtensions/PXContext.cs [hunk 15]
```diff
@@ -150,6 +143,33 @@ public FieldAttributesTypes(Compilation aCompilation)
             #endregion
         }
 		#endregion
+		#region Attributes Types
+		public class AttributesTypes
+		{
+			private readonly Compilation compilation;
```
Comment (dropsonic) [inline src/Acuminator/Acuminator.Utils/RoslynExtensions/PXContext.cs]: We agreed that underscores should be used for private members. See [this article](https://wiki.acumatica.com/display/ENG/Acuminator+team+code+style).

File: src/Acuminator/Acuminator.Utils/RoslynExtensions/PXFieldAttributes/AttributeInformation.cs [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+			var aggregateAttribute = _context.AttributeTypes.PXAggregateAttribute;
+			var dynamicAggregateAttribute = _context.AttributeTypes.PXDynamicAggregateAttribute;
+
+			if (ContainsBaseType(attributeSymbol, aggregateAttribute) || ContainsBaseType(attributeSymbol, dynamicAggregateAttribute))
+			{
+				var allAttributes = attributeSymbol.GetAllAttributesDefinedOnThisAndBaseTypes().ToList();
+				foreach (var attribute in allAttributes)
+				{
+					//go in recursuion
+					var result = AttributeDerivedFromClass(attribute, type, --depth);
```
[the remark this answers is not in the document]
Comment (dropsonic) [reply src/Acuminator/Acuminator.Utils/RoslynExtensions/PXFieldAttributes/AttributeInformation.cs]: @SENya1990 recursive algorithm is easier to understand; expected recursion level there is no more than 5-10 so I don't think there would be a performance hit.

File: src/Acuminator/Acuminator.Utils/RoslynExtensions/PXFieldAttributes/AttributeInformation.cs (hunk 4 above)
Comment (dropsonic) [reply src/Acuminator/Acuminator.Utils/RoslynExtensions/PXFieldAttributes/AttributeInformation.cs]: @SENya1990 I think that this is a mistake in Acumatica code. `PXDBScalarAttribute` is similar to `PXDBCalcedAttribute`, but it inherits from `PXDBFieldAttribute` and it is marked with `[PXAttributeFamily(typeof(PXDBFieldAttribute))]`.
I suggest to check if an attribute inherits from either `PXDBFieldAttribute` or `PXDBCalcedAttribute`. I don't think there is another way to determine it.

File: src/Samples/PX.Objects.HackathonDemo/PX.Objects.HackathonDemo/IIGPOALCLandedCost.cs [hunk 17]
```diff
@@ -0,0 +1,41 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using PX.Data;
+
+namespace PX.Objects.HackathonDemo
+{
+	[PXPrimaryGraph(typeof(IIGPOALCLandedCostEntry)), PXCacheName("Containers")]
```
Comment (dropsonic) [inline src/Samples/PX.Objects.HackathonDemo/PX.Objects.HackathonDemo/IIGPOALCLandedCost.cs]: This detail is not related to the diagnostic, so it'd better to remove it.

File: src/Acuminator/Acuminator.Analyzers/Analyzers/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeAnalyzer.cs (hunk 6 above)
Comment (SENya1990) [reply src/Acuminator/Acuminator.Analyzers/Analyzers/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeAnalyzer.cs]: @dropsonic, this was probably copied from field attributes analyzer, the reason behind try/catch is to intercept possible errors because I don't know if GetSyntaxAsync could throw exception. The OperationCancelled exception was added here to distinguidh between cancel and error - in case of error we probably should log it in the future

File: src/Acuminator/Acuminator.Utils/RoslynExtensions/PXFieldAttributes/AttributeInformation.cs (hunk 4 above)
Comment (SENya1990) [reply src/Acuminator/Acuminator.Utils/RoslynExtensions/PXFieldAttributes/AttributeInformation.cs]: @dropsonic actually there is a register of attributes which stores sets of bound and unbound field attributes, It can be used here, for code re-use. I wrote a small utility which retrieved attributes declared in Acumatica and added field attributes to register.  The algorithm checks if register contains any type from attribute's inheritance chain. 
Therefore, because bound field attributes set contains PXDBFieldAttribute, all attributes derived from it will be recognised as bound.

File: src/Acuminator/Acuminator.Utils/RoslynExtensions/PXFieldAttributes/AttributeInformation.cs [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+using System.Threading.Tasks;
+using Acuminator.Analyzers;
+using Microsoft.CodeAnalysis;
+
+namespace Acuminator.Utilities
+{
+	/// <summary>
+	/// Information about the Acumatica field attributes.
+	/// </summary>
+	public class AttributeInformation
```
Comment (SENya1990) [inline src/Acuminator/Acuminator.Utils/RoslynExtensions/PXFieldAttributes/AttributeInformation.cs]: @saloponov we already have helper FieldAttributesRegister. What is different in this helper?
Actually I would expect a different helper which from a given attribute/attribute collection will retrieve an expanded set of attributes including everything declared on aggregator attributes

File: src/Acuminator/Acuminator.Analyzers/Analyzers/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeAnalyzer.cs [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+
+			if (attributes.Length == 0 || symbolContext.CancellationToken.IsCancellationRequested)
+				return;
+
+			
+
+			if (symbolContext.CancellationToken.IsCancellationRequested)
+				return;
+			
+			bool isBoundField = attributeInformation.AreBoundAttributes(attributes.Select(a => a.AttributeClass).ToList());
```
Comment (SENya1990) [inline src/Acuminator/Acuminator.Analyzers/Analyzers/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeAnalyzer.cs]: @saloponov  please add more empty lines to separate control flow statements

File: src/Acuminator/Acuminator.Analyzers/Analyzers/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeAnalyzer.cs [hunk 20]
```diff
@@ -0,0 +1,153 @@
...
+				{
+					var typesHierarchy = attribute.AttributeClass.GetBaseTypesAndThis();
+					if(typesHierarchy.Contains(pxContext.AttributeTypes.PXDefaultAttribute))
+					{
+						foreach (var argument in attribute.NamedArguments)
+						{
+							if (argument.Key.Contains("PersistingCheck") && (int)argument.Value.Value == (int)PXPersistingCheck.Nothing)
+								return;
+						}
+						Location[] locations = await Task.WhenAll(GetAttributeLocationAsync(attribute, symbolContext.CancellationToken));
```
Comment (SENya1990) [inline src/Acuminator/Acuminator.Analyzers/Analyzers/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeAnalyzer.cs]: @saloponov ConfigureAwait(false) must be used in all await expressions

File: src/Acuminator/Acuminator.Analyzers/Analyzers/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeAnalyzer.cs [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+				}
+			}
+			return;
+		}
+
+		private static bool CheckProperty(IPropertySymbol property, PXContext pxContext)
+		{
+			var parent = property?.ContainingType;
+
+			if (parent == null || (!parent.ImplementsInterface(pxContext.IBqlTableType) && !parent.InheritsFrom(pxContext.PXCacheExtensionType)))
```
[the remark this answers is not in the document]
Comment (SENya1990) [reply src/Acuminator/Acuminator.Analyzers/Analyzers/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeAnalyzer.cs]: @saloponov  Ok, I forgot about ATR-88

File: src/Acuminator/Acuminator.Analyzers/Analyzers/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeAnalyzer.cs (hunk 20 above)
Comment (dropsonic) [reply src/Acuminator/Acuminator.Analyzers/Analyzers/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeAnalyzer.cs]: @SENya1990 @saloponov it is totally correct for the analyzers & code fixes, but you should be aware that in other places `ConfigureAwait(false)` is not the best choice. See [this article](https://github.com/Microsoft/vs-threading/blob/master/doc/cookbook_vs.md) for more details.

File: src/Acuminator/Acuminator.Analyzers/FixProviders/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeFix.cs [hunk 22]
```diff
@@ -0,0 +1,138 @@
...
+				modifiedRoot = root.ReplaceNode(attributeNode, pxUnboundDefaultAttribute.Attributes[0]);
+			}
+
+			return document.WithSyntaxRoot(modifiedRoot);
+		}
+
+
+	}
+
+	internal static class DiagnosticHelper {
```
Comment (SENya1990) [inline src/Acuminator/Acuminator.Analyzers/FixProviders/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeFix.cs]: @saloponov it is not neccessary to create separate class just for this method. Just simple method in the DacExtensionDefaultAttributeFix is enough:
```
private static bool IsBoundField(Diagnostic diagnostic)
{
	diagnostic.ThrowOnNull(nameof(diagnostic));
	return diagnostic.Properties.TryGetValue(DiagnosticProperty.IsBoundField, out string boundFlag)
				? boundFlag == bool.TrueString
			        :  false;
}
```
And it is used like a normal method

File: src/Acuminator/Acuminator.Analyzers/FixProviders/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeFix.cs [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+			
+			SyntaxNode modifiedRoot;
+
+			if (isBoundField)
+			{
+				if (attributeNode.ArgumentList != null)
+				{
+					var persistingCheckArguments = from arguments in attributeNode.ArgumentList.Arguments
+													where arguments.GetText().ToString().Contains(_PersistingCheck)
+													select arguments;
```
Comment (SENya1990) [inline src/Acuminator/Acuminator.Analyzers/FixProviders/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeFix.cs]: @dropsonic I'm not sure that the plain text check instead of syntax check should be used here. What do you think?

**Comment by SENya1990** [comment]
@saloponov AttributeInformation helper wasn't changed according to code review

File: src/Acuminator/Acuminator.Analyzers/FixProviders/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeFix.cs (hunk 23 above)
Comment (dropsonic) [reply src/Acuminator/Acuminator.Analyzers/FixProviders/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeFix.cs]: @saloponov it there a reason for plain text check?
@SENya1990 I think that all checks should be performed on a syntax tree

File: src/Acuminator/Acuminator.Analyzers/Analyzers/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeAnalyzer.cs [hunk 24]
```diff
@@ -0,0 +1,170 @@
...
+							Diagnostic.Create(
+								Descriptors.PX1030_DefaultAttibuteToExisitingRecords, attributeLocation, diagnosticProperties));
+					}
+				}
+			}
+		}
+
+		private static  bool isAttributeContainsPersistingCheckNothing(KeyValuePair<string, TypedConstant> argument)
+		{
+			return (argument.Key.Contains(_PersistingCheck) && (int)argument.Value.Value == (int)PXPersistingCheck.Nothing);
```
[the remark this answers is not in the document]
Comment (saloponov) [reply src/Acuminator/Acuminator.Analyzers/Analyzers/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeAnalyzer.cs]: Because in watch window expression ``` '(PXPersistingCheck)argument.Value.Value'``` threw an exception of type ```'System.InvalidCastException'```.

[the remark this answers is not in the document]
Comment (SENya1990) [reply src/Acuminator/Acuminator.Analyzers/Analyzers/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeAnalyzer.cs]: @saloponov and what is the type of argument.Value.Value? Is it just int? If so, then the existing code is ok

[the remark this answers is not in the document]
Comment (saloponov) [reply src/Acuminator/Acuminator.Analyzers/Analyzers/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeAnalyzer.cs]: Yes, argument.Value.Value is `int`

File: src/Acuminator/Acuminator.Utils/RoslynExtensions/PXFieldAttributes/AttributeInformation.cs [hunk 25]
```diff
@@ -0,0 +1,87 @@
...
+
+
+		public AttributeInformation(PXContext pxContext)
+		{
+			pxContext.ThrowOnNull(nameof(pxContext));
+
+			_context = pxContext;
+		}
+
+		public bool ContainsBaseType(ITypeSymbol attributeSymbol, ITypeSymbol type)
```
[the remark this answers is not in the document]
Comment (saloponov) [reply src/Acuminator/Acuminator.Utils/RoslynExtensions/PXFieldAttributes/AttributeInformation.cs]: @SENya1990 I going to use InheritsFromOrEquals(), is it?

[the remark this answers is not in the document]
Comment (SENya1990) [reply src/Acuminator/Acuminator.Utils/RoslynExtensions/PXFieldAttributes/AttributeInformation.cs]: @saloponov, I don't remember the exact func name, there should be one or two with the appropriate implementation, see for yourself

File: src/Acuminator/Acuminator.Analyzers/Analyzers/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeAnalyzer.cs [hunk 26]
```diff
@@ -0,0 +1,164 @@
...
+	{
+
+		private const string _PersistingCheck = "PersistingCheck";
+		
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+			ImmutableArray.Create
+			(
+				Descriptors.PX1030_DefaultAttibuteToExisitingRecords
+			);
+#pragma warning disable CS4014
```
Comment (dropsonic) [inline src/Acuminator/Acuminator.Analyzers/Analyzers/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeAnalyzer.cs]: Why?

File: src/Acuminator/Acuminator.Analyzers/Analyzers/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeAnalyzer.cs [hunk 27]
```diff
@@ -0,0 +1,164 @@
...
+		public static async Task<Location> GetAttributeLocationAsync(AttributeData attribute, CancellationToken cancellationToken)
+		{
+			SyntaxNode attributeSyntaxNode = null;
+
+			try
+			{
+				attributeSyntaxNode = await attribute.ApplicationSyntaxReference.GetSyntaxAsync(cancellationToken)
+																				.ConfigureAwait(false);
+			}
+			catch (OperationCanceledException)
```
Comment (dropsonic) [inline src/Acuminator/Acuminator.Analyzers/Analyzers/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeAnalyzer.cs]: Still try/catch here, please get rid of it

File: src/Acuminator/Acuminator.Tests/Tests/Unit/AttributeInformationTests.cs [hunk 28]
```diff
@@ -0,0 +1,286 @@
...
+
+		[Theory]
+		[EmbeddedFileData(@"Dac\PX1030\Unit\AggregateRecursiveAttributeInformationTest.cs")]
+		public void TestAggregateRegursiveAttribute(string source) =>
+			TestAttributeInformation(source, new List<bool> { true, false });
+		
+		private void TestAttributeInformation(string source, List<bool> expected)
+		{
+			Document document = CreateDocument(source);
+			SemanticModel semanticModel = document.GetSemanticModelAsync().Result;
```
Comment (dropsonic) [inline src/Acuminator/Acuminator.Tests/Tests/Unit/AttributeInformationTests.cs]: OK for now, but in the future please use `async` test cases and `await` instead of `Task.Result`

File: src/Acuminator/Acuminator.Analyzers/Analyzers/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeAnalyzer.cs [hunk 29]
```diff
@@ -0,0 +1,164 @@
...
+		
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+			ImmutableArray.Create
+			(
+				Descriptors.PX1030_DefaultAttibuteToExisitingRecords
+			);
+#pragma warning disable CS4014
+		internal override void AnalyzeCompilation(CompilationStartAnalysisContext compilationStartContext, PXContext pxContext)
+		{
+			compilationStartContext.RegisterSymbolAction(symbolContext =>
```
Comment (dropsonic) [inline src/Acuminator/Acuminator.Analyzers/Analyzers/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeAnalyzer.cs]: `async symbolContext => await AnalyzePropertyAsync`

Comment (SENya1990) [reply src/Acuminator/Acuminator.Analyzers/Analyzers/DAC/DacPropertyAttributes/DacExtensionDefaultAttributeAnalyzer.cs]: +1, I think I misused it too in DAC property / attributes type mismatch diagnostic so I'll fix it there

File: src/Acuminator/Acuminator.Utils/RoslynExtensions/PXFieldAttributes/AttributeInformation.cs [hunk 30]
```diff
@@ -0,0 +1,158 @@
...
+
+namespace Acuminator.Utilities
+{
+	/// <summary>
+	/// Information about the Acumatica field attributes.
+	/// </summary>
+	public class AttributeInformation
+	{
+		private readonly PXContext _context;
+		private static readonly string _attributeUsageAttribute = "AttributeUsageAttribute";
```
[the remark this answers is not in the document]
Comment (saloponov) [reply src/Acuminator/Acuminator.Utils/RoslynExtensions/PXFieldAttributes/AttributeInformation.cs]: @SENya1990 condition '''if (attribute.Name.Equals(_attributeUsageAttribute))''' cuts system attributes from list.

[the remark this answers is not in the document]
Comment (saloponov) [reply src/Acuminator/Acuminator.Utils/RoslynExtensions/PXFieldAttributes/AttributeInformation.cs]: @SENya1990 @dropsonic  `AttributeUsageAttribute` contains in attributes list defined on `PXAggregateAttribute` and get by `attributeSymbol.GetAllAttributesDefinedOnThisAndBaseTypes()`. This list doesn't contain `PXEventSubscriberAttribute`

File: src/Acuminator/Acuminator.Utils/RoslynExtensions/PXFieldAttributes/AttributeInformation.cs [hunk 31]
```diff
@@ -0,0 +1,155 @@
...
+
+						if (result)
+							return result;
+					}
+				}
+				return false;
+			}
+		}
+
+		public bool IsBoundAttribute(ITypeSymbol attributeSymbol)
```
Comment (dropsonic) [inline src/Acuminator/Acuminator.Utils/RoslynExtensions/PXFieldAttributes/AttributeInformation.cs]: It is better to have a register of base DB attributes (`PXDBFieldAttribute`, `PXDBCalcedAttribute`) and iterate through it because it can expand in the future.
